### PR TITLE
Integration with credentials plugin

### PR DESCRIPTION
**Pull request by Vlatombe**
This pull request provides integration with credentials plugin to replace framework principal/secret.

Credentials plugin provides a central facility to manage credentials and to update them only once if needed.

The motivation for this change is to provide the ability to update programmatically the mesos credentials without touching the cloud configuration, which may have been altered by an administrator. The current implementation doesn't allow that in an easy way, as you need to post the whole configuration.

### CODE REVIEW

**Comment by Vlatombe** [comment]
@reviewbybees

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 1]
```diff
@@ -417,6 +504,17 @@ public String getDisplayName() {
...
+    @SuppressWarnings("unused") // by stapler
+    public ListBoxModel doFillCredentialsIdItems(@AncestorInPath Item item, @QueryParameter String master) {
+      List<DomainRequirement> domainRequirements = (master == null) ? Collections.<DomainRequirement>emptyList()
+              : URIRequirementBuilder.fromUri(master.trim()).build();
+      return new StandardListBoxModel().withEmptySelection().withMatching(
+              CredentialsMatchers.instanceOf(UsernamePasswordCredentials.class),
+              CredentialsProvider.lookupCredentials(StandardUsernamePasswordCredentials.class, item, null,
+                      domainRequirements));
+    }
+
```
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: Are you using 4 spaces or 2 spaces? It is mixed in this plugin.

Comment (Vlatombe) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: Should be 2 spaces, I fixed occurrences in MesosCloud.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 2]
```diff
@@ -58,8 +76,11 @@
   private String description;
   private String frameworkName;
   private String slavesUser;
-  private String principal;
-  private String secret;
+  private String credentialsId;
+  @Deprecated
+  private transient String principal;
+  @Deprecated
+  private transient String secret;
```
Comment (tfennelly) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: Looking at the `readResolve` method, it seems `principal` and `secret` get reset to `null` after being used to resolve a `credentialsId`. Is there a need to track calls to the setters methods for these properties and then run the equivalent of the `readResolve` again? I assume the fact that the setter methods exist may mean someone can call them.

Comment (Vlatombe) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: I don't think I want to track these.
* I don't think the setters are being used at the moment
* I'd rather change setters implementations to throw Exceptions to prevent anyone using them, or even delete the methods to force potential client to use the credentials management instead.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java (hunk 1 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: @Vlatombe Great!

**Comment by Vlatombe** [comment]
@vinodkone ping

**Comment by stephenc** [comment]
:bee:

**Comment by tfennelly** [comment]
:bee: 

Assuming the build failure is expected/flaky.

**Comment by Vlatombe** [comment]
@reviewbybees done

File: src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/config.jelly [hunk 3]
```diff
@@ -19,12 +19,8 @@
         <f:textbox field="slavesUser" default=""/>
     </f:entry>
 
-    <f:entry title="${%Framework Principal}" field="principal">
-      <f:textbox field="principal" default="jenkins" />
-    </f:entry>
-
-    <f:entry title="${%Framework Secret}" field="secret" description="${%Specify secret to enable framework authentication}">
-      <f:textbox field="secret"/>
+    <f:entry field="credentialsId" title="${%Framework credentials}">
```
Comment (vinodkone) [inline src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/config.jelly]: Is this a backwards compatible change? If not, we can just remove these fields.

File: src/test/java/org/jenkinsci/plugins/mesos/MesosCloudTest.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+import static org.junit.Assert.assertTrue;
+
+public class MesosCloudTest {
+
+    @Rule
+    public JenkinsRule r = new JenkinsRule();
+
+    /**
+     * Ignored because I'm getting Cannot call method "hasClassName" of undefined (http://localhost:55154/jenkins/adjuncts/b4f04185/lib/form/hetero-list/hetero-list.js#16)
```
Comment (vinodkone) [inline src/test/java/org/jenkinsci/plugins/mesos/MesosCloudTest.java]: Please fix this test. Adding a test that doesn't work sets a bad precedent.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 5]
```diff
@@ -417,6 +507,17 @@ public String getDisplayName() {
       return "Mesos Cloud";
     }
 
+    @Restricted(DoNotUse.class) // stapler only
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: We use fully formed sentences for our comments, i.e., we use proper capitalization and periods.

// Stapler only.

Please fix and everywhere else in this patch.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 6]
```diff
@@ -390,6 +453,33 @@ public JSONObject getSlaveAttributeForLabel(String labelName) {
     return null;
   }
 
+  protected Object readResolve() {
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: It's really hard to understand what's happening in this method. Even the name of the method doesn't give me a strong clue. This could use some comments.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 7]
```diff
@@ -208,6 +229,16 @@ public void restartMesos() {
 
   }
 
+  public StandardUsernamePasswordCredentials getCredentials() {
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: Add some comments please.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 8]
```diff
@@ -58,8 +76,11 @@
   private String description;
   private String frameworkName;
   private String slavesUser;
-  private String principal;
-  private String secret;
+  private String credentialsId;
+  @Deprecated
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: We cannot deprecate these without giving a heads up to existing users. Can you please email our mailing list and get feedback about this change?

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 9]
```diff
@@ -14,15 +14,27 @@
  */
 package org.jenkinsci.plugins.mesos;
 
+import com.cloudbees.plugins.credentials.CredentialsMatchers;
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: I'm assuming these packages are all open source without any licensing conflicts?

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 10]
```diff
@@ -14,15 +14,27 @@
...
+import com.cloudbees.plugins.credentials.CredentialsProvider;
+import com.cloudbees.plugins.credentials.CredentialsScope;
+import com.cloudbees.plugins.credentials.SystemCredentialsProvider;
+import com.cloudbees.plugins.credentials.common.StandardListBoxModel;
+import com.cloudbees.plugins.credentials.common.StandardUsernamePasswordCredentials;
+import com.cloudbees.plugins.credentials.common.UsernamePasswordCredentials;
+import com.cloudbees.plugins.credentials.domains.DomainRequirement;
+import com.cloudbees.plugins.credentials.domains.URIRequirementBuilder;
+import com.cloudbees.plugins.credentials.impl.UsernamePasswordCredentialsImpl;
 import hudson.Extension;
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: new line.

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java [hunk 11]
```diff
@@ -114,42 +116,49 @@ public synchronized void init() {
...
             driver = new MesosSchedulerDriver(JenkinsScheduler.this, framework, mesosCloud.getMaster(), credential);
-        } else {
+          } else {
             driver = new MesosSchedulerDriver(JenkinsScheduler.this, framework, mesosCloud.getMaster());
+          }
+          Status runStatus = driver.run();
+          if (runStatus != Status.DRIVER_STOPPED) {
+            LOGGER.severe("The Mesos driver was aborted! Status code: " + runStatus.getNumber());
+          }
+        } catch (RuntimeException e) {
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: why is a "try catch" changed sneaked in here? If it's not related to this change, please split that out into its own PR.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java (hunk 6 above)
Comment (Vlatombe) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: This is a common Jenkins plugin practice to retain backward compatibility, but at the same time migrate the old data to the new data structure. See https://wiki.jenkins-ci.org/display/JENKINS/Hint+on+retaining+backward+compatibility for details. I will add comments on the implementation details.

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java [hunk 12]
```diff
@@ -114,42 +116,49 @@ public synchronized void init() {
...
-        + "\n" + "Checkpointing: " + framework.getCheckpoint()
-        );
-
-        if (StringUtils.isNotBlank(mesosCloud.getSecret())) {
+        try {
+          String targetUser = mesosCloud.getSlavesUser();
+          String webUrl = Jenkins.getInstance().getRootUrl();
+          if (webUrl == null) webUrl = System.getenv("JENKINS_URL");
+          StandardUsernamePasswordCredentials credentials = mesosCloud.getCredentials();
+          String principal = credentials == null ? "" : credentials.getUsername();
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: Previously, the default principal was "jenkins". Now it is being changed to empty string. This is not a backwards compatible change!

**Comment by vinodkone** [comment]
There are some compatibility concerns with this change! Lets fix those.

More importantly, I don't understand the motivation for this change. Can you please include that in the description?

We definitely need feedback from the community regarding this change. Please send an email to the mailing list with the motivation and link to the PR.

Thanks.

File: src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/config.jelly (hunk 3 above)
Comment (Vlatombe) [inline src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/config.jelly]: I retain the fields as a matter of backward compatibility, otherwise it is much more complicated code-wise to migrate the existing data.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java (hunk 9 above)
Comment (Vlatombe) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: https://github.com/jenkinsci/credentials-plugin/blob/master/src/main/java/com/cloudbees/plugins/credentials/CredentialsMatchers.java

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java (hunk 11 above)
Comment (Vlatombe) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: If any exception is raised during the thread execution, the thread just dies without any log. And worse, as the lines
```
driver = null;
running = false;
```
were not in a finally block, the scheduler would still be considered as running. Just set your authentication wrong once and you end up stuck.

**Comment by Vlatombe** [comment]
Hi,

I think I have addressed your review comments, and I offloaded some items to other pull requests as requested.

**Comment by brism** [comment]
:+1: 

If I understand this change correctly, it would solve a related issue I have with using the plugin in our environment. 

We're standing up a Jenkins master for any dev team that wants one, and they'll run slave jobs on Mesos. I would like to be able to give these teams Admin to configure the masters as they see fit.

The problem is that the **secret** for the Mesos cluster is right there *in plain text* on the Jenkins Config page. From a security perspective it would be much better if this were handled like a proper credential. I currently can't let them see this page or they'd have the creds for the Mesos cluster.

Thanks for the good work!
